src/examples/AppUartModule.cpp

```diff
@@ -0,0 +1,270 @@
+#endif
+                lineToReadAvailable = true;
+                GS->terminal.CheckAndProcessLine();
+                lineToReadAvailable = false;
+                readBufferOffset = 0;
+            }
+        }
+    }
+}
+
+NodeId AppUartModule::GetMeshAccessConnectionInNodeId(bool isVirtual) const
+{
+    MeshAccessConnections connIn = GS->cm.GetMeshAccessConnections(ConnectionDirection::DIRECTION_IN);
+    if (connIn.count < 1)
+        return NODE_ID_INVALID;
+    return isVirtual ? connIn.handles[0].GetVirtualPartnerId() : connIn.handles[0].GetPartnerId();
+}
+
+bool AppUartModule::IsConnectSmartPhone()
+{
+    const NodeId partnerId = GetMeshAccessConnectionInNodeId();
+    return partnerId >= NODE_ID_APP_BASE && partnerId < NODE_ID_GLOBAL_DEVICE_BASE;
+}
+
+void AppUartModule::SendAppLogQueue()
+{
+    if (logQueue._numElements < 2 || !IsConnectSmartPhone())
+        return;
+
+    AppUartLogRemain* logRemain = reinterpret_cast<AppUartLogRemain*>(logQueue.PeekNext().data);
+    char* log = reinterpret_cast<char*>(logQueue.PeekNext(1).data);
+    const u16 logLen = logRemain->logLen;
+    AppUartModuleTerminalLogMessage message;
+    if (logLen <= DATA_MAX_LEN) {
+        message.splitHeader = MessageType::SPLIT_WRITE_CMD_END;
+        message.splitCount = 0;
+        message.partLen = logLen;
+        CheckedMemcpy(message.data, log, logLen);
+    }
+    const u16 remainLogLen = logLen - logRemain->sentLength;
+    if (logLen > DATA_MAX_LEN) {
+        message.splitHeader
+            = remainLogLen > DATA_MAX_LEN ? MessageType::SPLIT_WRITE_CMD : MessageType::SPLIT_WRITE_CMD_END;
+        message.splitCount = (u8)(logRemain->sentLength / DATA_MAX_LEN);
+        message.partLen = remainLogLen > DATA_MAX_LEN ? DATA_MAX_LEN : remainLogLen;
+        CheckedMemcpy(message.data, &log[logRemain->sentLength], message.partLen);
+    }
+
+    const NodeId virtualPartnerId = GetMeshAccessConnectionInNodeId(true);
+    SendModuleActionMessage(MessageType::MODULE_ACTION_RESPONSE, virtualPartnerId,
+        AppUartModuleActionResponseMessages::RECEIVE_LOG, 0, reinterpret_cast<u8*>(&message),
+        SIZEOF_APP_UART_MODULE_TERMINAL_COMMAND_MESSAGE_STATIC + message.partLen, false);
+    if (remainLogLen > DATA_MAX_LEN) {
+        logRemain->sentLength += DATA_MAX_LEN;
+        return;
+    }
+    // send log completely
+    DiscardLogQueue();
+}
+
+bool AppUartModule::PutAppLogQueue(const char* log, u16 length)
+{
+    if (log == NULL || length == 0 || !IsConnectSmartPhone())
+        return false;
+
+    AppUartLogRemain remain = { .sentLength = 0, .logLen = length };
+    if (!logQueue.Put(reinterpret_cast<u8*>(&remain), sizeof(AppUartLogRemain)))
+        return false;
+    if (!logQueue.Put(reinterpret_cast<u8*>(const_cast<char*>(log)), length)) {
+        logQueue.DiscardLast();
+        return false;
+    }
+    return true;
+}
+
+void AppUartModule::DiscardLogQueue()
+{
+    logQueue.DiscardNext();
```

Comment:
This method is not used to clear the queue, but to delete the log from the queue when one log is successfully sent.
Since the log is stored in the queue in two areas, the sent length and the actual log, in case the log exceeds the maximum send size, `DiscardNext()` is called twice in case of a successful send.

Reply:
Perfect, thank you :+1: